Piped input does not respect excluded paths

**Description of the issue**

PSRule does not respect exluded paths in `ps-rule.yaml`

**To Reproduce**

```
input:
  pathIgnore:
    # Ignore common files that don't need analysis.
    - "**/bicepconfig.json"
    - "*.md"
    - "*.png"
    - ".azuredevops/"

    # Exclude Bicep module files.
    - "modules/**/*.bicep"

    # Exclude JSON module files.
    - "modules/**/*.json"

    # Include test/main.test.bicep files from modules.
    - "!modules/**/test/*.bicep"
```
I also have checked ps-rule branch in ALZ-Bicep and I can see it's solved the same way there?

Also tried 

`- "!/modules/**/*.test.bicep"`

The task

```
Install-Module PSRule.Rules.Azure -Scope CurrentUser -Force;
git diff --diff-filter=d HEAD^ HEAD --name-only | ForEach-Object { Get-ChildItem $_ } | Assert-PSRule -Format File -Module 'PSRule.Rules.Azure' -Baseline 'Azure.GA_2022_03' -OutputFormat NUnit3 -OutputPath 'reports/ps-rule-results.xml'
```

**Expected behaviour**

Should only traverse test folder with it's underlaying .bicep content.

**Error output**

```
2022-06-17T11:43:47.8240390Z ##[error]Failed to expand bicep source '/home/vsts/work/1/s/modules/authorization/roleassignments-automation-account/main.bicep'. Exception calling "GetBicepResources" with "3" argument(s): "Unable to expand resources because the source file '/home/vsts/work/1/s/modules/authorization/roleassignments-automation-account/main.bicep' was not valid. An error occurred evaluating expression '[length(parameters('parPrincipalIds'))]' line 92. The parameter named 'parPrincipalIds' was not set or a defaultValue was defined."
2022-06-17T11:43:50.8662218Z 
2022-06-17T11:43:50.8664516Z  -> /home/vsts/work/1/s/modules/authorization/roleassignments-automation-account/test/main.test.bicep : Microsoft.Resources/deployments [2/2]
2022-06-17T11:43:50.8668930Z 
2022-06-17T11:43:50.8675905Z     [PASS] Azure.Template.ParameterStrongType (AZR-000227)
2022-06-17T11:43:50.8679748Z     [PASS] Azure.Template.ExpressionLength (AZR-000228)
2022-06-17T11:43:50.8905808Z 
2022-06-17T11:43:50.8908502Z  -> /home/vsts/work/1/s/modules/authorization/roleassignments-automation-account/test/main.test.bicep : Microsoft.Resources/deployments [2/2]
2022-06-17T11:43:50.8909925Z 
2022-06-17T11:43:50.8911413Z     [PASS] Azure.Template.ParameterStrongType (AZR-000227)
2022-06-17T11:43:50.8912198Z     [PASS] Azure.Template.ExpressionLength (AZR-000228)
```
As seen it's also trying to traverse thru main.bicep which is not buildable. 

**Module in use and version:**

- Module: PSRule
- Version: v2.1.0

I am quite unsure if this is a bug or just human error from my side.

--- a/README.md
+++ b/README.md
@@ -252,6 +252,7 @@ The following conceptual topics exist in the `PSRule` module:
   - [Include.Path](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#includepath)
   - [Input.Format](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputformat)
   - [Input.IgnoreGitPath](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignoregitpath)
+  - [Input.IgnoreObjectSource](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignoreobjectsource)
   - [Input.IgnoreRepositoryCommon](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignorerepositorycommon)
   - [Input.ObjectPath](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputobjectpath)
   - [Input.PathIgnore](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputpathignore)

--- a/docs/CHANGELOG-v2.md
+++ b/docs/CHANGELOG-v2.md
@@ -18,6 +18,8 @@ What's changed since pre-release v2.2.0-B0021:
 - Bug fixes:
   - Fixed output of reason with wide format by @BernieWhite.
     [#1117](https://github.com/microsoft/PSRule/issues/1117)
+  - Fixed piped input does not respect excluded paths by @BernieWhite.
+    [#1114](https://github.com/microsoft/PSRule/issues/1114)
 
 ## v2.2.0-B0021 (pre-release)
 

--- a/docs/concepts/PSRule/en-US/about_PSRule_Options.md
+++ b/docs/concepts/PSRule/en-US/about_PSRule_Options.md
@@ -24,6 +24,7 @@ The following workspace options are available for use:
 - [Include.Path](#includepath)
 - [Input.Format](#inputformat)
 - [Input.IgnoreGitPath](#inputignoregitpath)
+- [Input.IgnoreObjectSource](#inputignoreobjectsource)
 - [Input.IgnoreRepositoryCommon](#inputignorerepositorycommon)
 - [Input.ObjectPath](#inputobjectpath)
 - [Input.PathIgnore](#inputpathignore)
@@ -1206,6 +1207,55 @@ env:
 variables:
 - name: PSRULE_INPUT_IGNOREGITPATH
   value: false
+```
+
+### Input.IgnoreObjectSource
+
+By default, objects read from file using `inputPath` will be skipped if the file path has been ignored.
+When set to true, additionally objects with a source path that has been ignored will be skipped.
+This will include `FileInfo` objects, and objects with a source set using the `_PSRule.source` property.
+
+File paths to ignore are set by `Input.PathIgnore`, `Input.IgnoreGitPath`, and `Input.IgnoreRepositoryCommon`.
+
+This option can be specified using:
+
+```powershell
+# PowerShell: Using the InputIgnoreObjectSource parameter
+$option = New-PSRuleOption -InputIgnoreObjectSource $True;
+```
+
+```powershell
+# PowerShell: Using the Input.IgnoreObjectSource hashtable key
+$option = New-PSRuleOption -Option @{ 'Input.IgnoreObjectSource' = $True };
+```
+
+```powershell
+# PowerShell: Using the InputIgnoreObjectSource parameter to set YAML
+Set-PSRuleOption -InputIgnoreObjectSource $True;
+```
+
+```yaml
+# YAML: Using the input/ignoreObjectSource property
+input:
+  ignoreObjectSource: true
+```
+
+```bash
+# Bash: Using environment variable
+export PSRULE_INPUT_IGNOREOBJECTSOURCE=true
+```
+
+```yaml
+# GitHub Actions: Using environment variable
+env:
+  PSRULE_INPUT_IGNOREOBJECTSOURCE: true
+```
+
+```yaml
+# Azure Pipelines: Using environment variable
+variables:
+- name: PSRULE_INPUT_IGNOREOBJECTSOURCE
+  value: true
 ```
 
 ### Input.IgnoreRepositoryCommon
@@ -2638,6 +2688,8 @@ include:
 input:
   format: Yaml
   ignoreGitPath: false
+  ignoreObjectSource: true
+  ignoreRepositoryCommon: false
   objectPath: items
   pathIgnore:
   - '*.Designer.cs'
@@ -2743,6 +2795,8 @@ include:
 input:
   format: Detect
   ignoreGitPath: true
+  ignoreObjectSource: false
+  ignoreRepositoryCommon: true
   objectPath: null
   pathIgnore: [ ]
   targetType: [ ]

--- a/schemas/PSRule-options.schema.json
+++ b/schemas/PSRule-options.schema.json
@@ -198,6 +198,7 @@
             "type": "object",
             "title": "Input options",
             "description": "Options that affect how input types are processed.",
+            "markdownDescription": "Options that affect how input types are processed.",
             "properties": {
                 "format": {
                     "type": "string",
@@ -218,16 +219,23 @@
                 "ignoreGitPath": {
                     "type": "boolean",
                     "title": "Ignore .git path",
-                    "description": "Determine if files within the .git path are ignored when the -InputPath parameter is used.",
-                    "markdownDescription": "Determine if files within the `.git` path are ignored when the `-InputPath` parameter is used. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignoregitpath)",
+                    "description": "Determine if files within the .git path are ignored when the -InputPath parameter is used. This is enabled by default.",
+                    "markdownDescription": "Determine if files within the `.git` path are ignored when the `-InputPath` parameter is used. This is enabled by default. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignoregitpath)",
                     "default": true
                 },
                 "ignoreRepositoryCommon": {
                     "type": "boolean",
                     "title": "Ignore common files",
-                    "description": "Determine if common repository files are ignored when the -InputPath parameter is used.",
-                    "markdownDescription": "Determine if common repository files are ignored when the `-InputPath` parameter is used. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignorerepositorycommon)",
+                    "description": "Determine if common repository files are ignored when the -InputPath parameter is used. This is enabled by default.",
+                    "markdownDescription": "Determine if common repository files are ignored when the `-InputPath` parameter is used. This is enabled by default. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignorerepositorycommon)",
                     "default": true
+                },
+                "ignoreObjectSource": {
+                    "type": "boolean",
+                    "title": "Ignore object source",
+                    "description": "Determines if objects are ignore based on their file source path. When set, objects from the pipeline or read from files will be exclude based on their source path and the configuration of pathIgnore, ignoreGitPath, and ignoreRepositoryCommon options.",
+                    "markdownDescription": "Determines if objects are ignore based on their file source path. When set, objects from the pipeline or read from files will be exclude based on their source path and the configuration of `pathIgnore`, `ignoreGitPath`, and `ignoreRepositoryCommon` options. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputignoreobjectsource)",
+                    "default": false
                 },
                 "objectPath": {
                     "type": "string",
@@ -238,8 +246,8 @@
                 "pathIgnore": {
                     "type": "array",
                     "title": "Path ignore",
-                    "description": "Ignores input files that match the path spec.",
-                    "markdownDescription": "Ignores input files that match the path spec. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputpathignore)",
+                    "description": "Exclude input files that match the path spec. To exclude a file specify the path prefix or file name. To re-include previously excluded paths use ! at the start of the entry. Wildcards are supported.",
+                    "markdownDescription": "Exclude input files that match the path spec. To exclude a file specify the path prefix or file name. To re-include previously excluded paths use `!` at the start of the entry. Wildcards are supported. [See help](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#inputpathignore)",
                     "items": {
                         "type": "string"
                     },

--- a/src/PSRule/Common/PSObjectExtensions.cs
+++ b/src/PSRule/Common/PSObjectExtensions.cs
@@ -4,6 +4,7 @@
 using System;
 using System.Collections;
 using System.Globalization;
+using System.IO;
 using System.Management.Automation;
 using System.Threading;
 using Newtonsoft.Json;
@@ -135,6 +136,21 @@ namespace PSRule
                     var issue = TargetIssueInfo.Create(issues.GetValue(i));
                     targetInfo.WithIssue(issue);
                 }
+            }
+        }
+
+        public static void ConvertTargetInfoType(this PSObject o)
+        {
+            var info = o?.BaseObject;
+            if (info is FileInfo fileInfo)
+            {
+                UseTargetInfo(o, out var targetInfo);
+                targetInfo.WithSource(new TargetSourceInfo(fileInfo));
+            }
+            if (info is InputFileInfo inputFileInfo)
+            {
+                UseTargetInfo(o, out var targetInfo);
+                targetInfo.WithSource(new TargetSourceInfo(inputFileInfo));
             }
         }
 

--- a/src/PSRule/Configuration/InputOption.cs
+++ b/src/PSRule/Configuration/InputOption.cs
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 using System;
@@ -12,8 +12,9 @@ namespace PSRule.Configuration
     /// </summary>
     public sealed class InputOption : IEquatable<InputOption>
     {
-        private const InputFormat DEFAULT_FORMAT = PSRule.Configuration.InputFormat.Detect;
+        private const InputFormat DEFAULT_FORMAT = InputFormat.Detect;
         private const bool DEFAULT_IGNOREGITPATH = true;
+        private const bool DEFAULT_IGNOREOBJECTSOURCE = false;
         private const bool DEFAULT_IGNOREREPOSITORYCOMMON = true;
         private const string DEFAULT_OBJECTPATH = null;
         private const string[] DEFAULT_PATHIGNORE = null;
@@ -23,6 +24,7 @@ namespace PSRule.Configuration
         {
             Format = DEFAULT_FORMAT,
             IgnoreGitPath = DEFAULT_IGNOREGITPATH,
+            IgnoreObjectSource = DEFAULT_IGNOREOBJECTSOURCE,
             IgnoreRepositoryCommon = DEFAULT_IGNOREREPOSITORYCOMMON,
             ObjectPath = DEFAULT_OBJECTPATH,
             PathIgnore = DEFAULT_PATHIGNORE,
@@ -33,6 +35,7 @@ namespace PSRule.Configuration
         {
             Format = null;
             IgnoreGitPath = null;
+            IgnoreObjectSource = null;
             IgnoreRepositoryCommon = null;
             ObjectPath = null;
             PathIgnore = null;
@@ -46,6 +49,7 @@ namespace PSRule.Configuration
 
             Format = option.Format;
             IgnoreGitPath = option.IgnoreGitPath;
+            IgnoreObjectSource = option.IgnoreObjectSource;
             IgnoreRepositoryCommon = option.IgnoreRepositoryCommon;
             ObjectPath = option.ObjectPath;
             PathIgnore = option.PathIgnore;
@@ -62,6 +66,7 @@ namespace PSRule.Configuration
             return other != null &&
                 Format == other.Format &&
                 IgnoreGitPath == other.IgnoreGitPath &&
+                IgnoreObjectSource == other.IgnoreObjectSource &&
                 IgnoreRepositoryCommon == other.IgnoreRepositoryCommon &&
                 ObjectPath == other.ObjectPath &&
                 PathIgnore == other.PathIgnore &&
@@ -75,6 +80,7 @@ namespace PSRule.Configuration
                 var hash = 17;
                 hash = hash * 23 + (Format.HasValue ? Format.Value.GetHashCode() : 0);
                 hash = hash * 23 + (IgnoreGitPath.HasValue ? IgnoreGitPath.Value.GetHashCode() : 0);
+                hash = hash * 23 + (IgnoreObjectSource.HasValue ? IgnoreObjectSource.Value.GetHashCode() : 0);
                 hash = hash * 23 + (IgnoreRepositoryCommon.HasValue ? IgnoreRepositoryCommon.Value.GetHashCode() : 0);
                 hash = hash * 23 + (ObjectPath != null ? ObjectPath.GetHashCode() : 0);
                 hash = hash * 23 + (PathIgnore != null ? PathIgnore.GetHashCode() : 0);
@@ -89,6 +95,7 @@ namespace PSRule.Configuration
             {
                 Format = o1.Format ?? o2.Format,
                 IgnoreGitPath = o1.IgnoreGitPath ?? o2.IgnoreGitPath,
+                IgnoreObjectSource = o1.IgnoreObjectSource ?? o2.IgnoreObjectSource,
                 IgnoreRepositoryCommon = o1.IgnoreRepositoryCommon ?? o2.IgnoreRepositoryCommon,
                 ObjectPath = o1.ObjectPath ?? o2.ObjectPath,
                 PathIgnore = o1.PathIgnore ?? o2.PathIgnore,
@@ -108,6 +115,12 @@ namespace PSRule.Configuration
         /// </summary>
         [DefaultValue(null)]
         public bool? IgnoreGitPath { get; set; }
+
+        /// <summary>
+        /// Determines if objects are ignore based on their file source path.
+        /// </summary>
+        [DefaultValue(null)]
+        public bool? IgnoreObjectSource { get; set; }
 
         /// <summary>
         /// Determine if common repository files are ignored.
@@ -141,6 +154,9 @@ namespace PSRule.Configuration
             if (env.TryBool("PSRULE_INPUT_IGNOREGITPATH", out var ignoreGitPath))
                 IgnoreGitPath = ignoreGitPath;
 
+            if (env.TryBool("PSRULE_INPUT_IGNOREOBJECTSOURCE", out var ignoreObjectSource))
+                IgnoreObjectSource = ignoreObjectSource;
+
             if (env.TryBool("PSRULE_INPUT_IGNOREREPOSITORYCOMMON", out var ignoreRepositoryCommon))
                 IgnoreRepositoryCommon = ignoreRepositoryCommon;
 
@@ -161,6 +177,9 @@ namespace PSRule.Configuration
 
             if (index.TryPopBool("Input.IgnoreGitPath", out var ignoreGitPath))
                 IgnoreGitPath = ignoreGitPath;
+
+            if (index.TryPopBool("Input.IgnoreObjectSource", out var ignoreObjectSource))
+                IgnoreObjectSource = ignoreObjectSource;
 
             if (index.TryPopBool("Input.IgnoreRepositoryCommon", out var ignoreRepositoryCommon))
                 IgnoreRepositoryCommon = ignoreRepositoryCommon;

--- a/src/PSRule/Data/ITargetSourceCollection.cs
+++ b/src/PSRule/Data/ITargetSourceCollection.cs
@@ -18,6 +18,8 @@ namespace PSRule.Data
 
         internal TargetSourceCollection() { }
 
+        public int Count => _Items == null ? 0 : _Items.Count;
+
         public TargetSourceInfo this[string type]
         {
             get

--- a/src/PSRule/PSRule.psm1
+++ b/src/PSRule/PSRule.psm1
@@ -1203,6 +1203,10 @@ function New-PSRuleOption {
         [Parameter(Mandatory = $False)]
         [System.Boolean]$InputIgnoreRepositoryCommon = $True,
 
+        # Sets the Input.IgnoreObjectSource option
+        [Parameter(Mandatory = $False)]
+        [System.Boolean]$InputIgnoreObjectSource = $False,
+
         # Sets the Input.ObjectPath option
         [Parameter(Mandatory = $False)]
         [Alias('InputObjectPath')]
@@ -1468,6 +1472,10 @@ function Set-PSRuleOption {
         # Sets the Input.IgnoreGitPath option
         [Parameter(Mandatory = $False)]
         [System.Boolean]$InputIgnoreGitPath = $True,
+
+        # Sets the Input.IgnoreObjectSource option
+        [Parameter(Mandatory = $False)]
+        [System.Boolean]$InputIgnoreObjectSource = $False,
 
         # Sets the Input.IgnoreRepositoryCommon option
         [Parameter(Mandatory = $False)]
@@ -2181,6 +2189,10 @@ function SetOptions {
         [Parameter(Mandatory = $False)]
         [System.Boolean]$InputIgnoreGitPath = $True,
 
+        # Sets the Input.IgnoreObjectSource option
+        [Parameter(Mandatory = $False)]
+        [System.Boolean]$InputIgnoreObjectSource = $False,
+
         # Sets the Input.IgnoreRepositoryCommon option
         [Parameter(Mandatory = $False)]
         [System.Boolean]$InputIgnoreRepositoryCommon = $True,
@@ -2362,6 +2374,11 @@ function SetOptions {
         # Sets option Input.IgnoreGitPath
         if ($PSBoundParameters.ContainsKey('InputIgnoreGitPath')) {
             $Option.Input.IgnoreGitPath = $InputIgnoreGitPath;
+        }
+
+        # Sets option Input.IgnoreObjectSource
+        if ($PSBoundParameters.ContainsKey('InputIgnoreObjectSource')) {
+            $Option.Input.IgnoreObjectSource = $InputIgnoreObjectSource;
         }
 
         # Sets option Input.IgnoreRepositoryCommon

--- a/src/PSRule/Pipeline/GetTargetPipeline.cs
+++ b/src/PSRule/Pipeline/GetTargetPipeline.cs
@@ -47,14 +47,7 @@ namespace PSRule.Pipeline
             if (path == null || path.Length == 0)
                 return;
 
-            var basePath = PSRuleOption.GetWorkingPath();
-            var ignoreGitPath = Option.Input.IgnoreGitPath ?? InputOption.Default.IgnoreGitPath.Value;
-            var ignoreRepositoryCommon = Option.Input.IgnoreRepositoryCommon ?? InputOption.Default.IgnoreRepositoryCommon.Value;
-            var filter = PathFilterBuilder.Create(basePath, Option.Input.PathIgnore, ignoreGitPath, ignoreRepositoryCommon);
-            if (Option.Input.Format == InputFormat.File)
-                filter.UseGitIgnore();
-
-            var builder = new InputPathBuilder(GetOutput(), basePath, "*", filter.Build());
+            var builder = new InputPathBuilder(GetOutput(), PSRuleOption.GetWorkingPath(), "*", GetInputFilter());
             builder.Add(path);
             _InputPath = builder.Build();
         }
@@ -116,7 +109,7 @@ namespace PSRule.Pipeline
                     return PipelineReceiverActions.DetectInputFormat(sourceObject, next);
                 });
             }
-            return new PipelineReader(VisitTargetObject, _InputPath);
+            return new PipelineReader(VisitTargetObject, _InputPath, GetInputObjectSourceFilter());
         }
     }
 

--- a/src/PSRule/Pipeline/InvokeRulePipeline.cs
+++ b/src/PSRule/Pipeline/InvokeRulePipeline.cs
@@ -36,14 +36,7 @@ namespace PSRule.Pipeline
             if (path == null || path.Length == 0)
                 return;
 
-            var basePath = PSRuleOption.GetWorkingPath();
-            var ignoreGitPath = Option.Input.IgnoreGitPath ?? InputOption.Default.IgnoreGitPath.Value;
-            var ignoreRepositoryCommon = Option.Input.IgnoreRepositoryCommon ?? InputOption.Default.IgnoreRepositoryCommon.Value;
-            var filter = PathFilterBuilder.Create(basePath, Option.Input.PathIgnore, ignoreGitPath, ignoreRepositoryCommon);
-            if (Option.Input.Format == InputFormat.File)
-                filter.UseGitIgnore();
-
-            var builder = new InputPathBuilder(GetOutput(), basePath, "*", filter.Build());
+            var builder = new InputPathBuilder(GetOutput(), PSRuleOption.GetWorkingPath(), "*", GetInputFilter());
             builder.Add(path);
             _InputPath = builder.Build();
         }
@@ -150,7 +143,7 @@ namespace PSRule.Pipeline
                     return PipelineReceiverActions.DetectInputFormat(sourceObject, next);
                 });
             }
-            return new PipelineReader(VisitTargetObject, _InputPath);
+            return new PipelineReader(VisitTargetObject, _InputPath, GetInputObjectSourceFilter());
         }
     }
 

--- a/src/PSRule/Pipeline/PipelineBuilder.cs
+++ b/src/PSRule/Pipeline/PipelineBuilder.cs
@@ -124,6 +124,7 @@ namespace PSRule.Pipeline
         private Hashtable _Tag;
         private BaselineOption _Baseline;
         private string[] _Convention;
+        private PathFilter _InputFilter;
 
         private readonly HostPipelineWriter _Output;
 
@@ -280,7 +281,7 @@ namespace PSRule.Pipeline
 
         protected virtual PipelineReader PrepareReader()
         {
-            return new PipelineReader(null, null);
+            return new PipelineReader(null, null, GetInputObjectSourceFilter());
         }
 
         protected virtual PipelineWriter PrepareWriter()
@@ -339,6 +340,27 @@ namespace PSRule.Pipeline
                 result.AddRange(parent);
 
             return result.Count == 0 ? null : result.ToArray();
+        }
+
+        protected PathFilter GetInputObjectSourceFilter()
+        {
+            return Option.Input.IgnoreObjectSource.GetValueOrDefault(InputOption.Default.IgnoreObjectSource.Value) ? GetInputFilter() : null;
+        }
+
+        protected PathFilter GetInputFilter()
+        {
+            if (_InputFilter == null)
+            {
+                var basePath = PSRuleOption.GetWorkingPath();
+                var ignoreGitPath = Option.Input.IgnoreGitPath ?? InputOption.Default.IgnoreGitPath.Value;
+                var ignoreRepositoryCommon = Option.Input.IgnoreRepositoryCommon ?? InputOption.Default.IgnoreRepositoryCommon.Value;
+                var builder = PathFilterBuilder.Create(basePath, Option.Input.PathIgnore, ignoreGitPath, ignoreRepositoryCommon);
+                if (Option.Input.Format == InputFormat.File)
+                    builder.UseGitIgnore();
+
+                _InputFilter = builder.Build();
+            }
+            return _InputFilter;
         }
 
         private OptionContext GetOptionContext()

--- a/src/PSRule/Pipeline/PipelineReader.cs
+++ b/src/PSRule/Pipeline/PipelineReader.cs
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 using System;
@@ -12,12 +12,14 @@ namespace PSRule.Pipeline
     {
         private readonly VisitTargetObject _Input;
         private readonly InputFileInfo[] _InputPath;
+        private readonly PathFilter _InputFilter;
         private readonly ConcurrentQueue<TargetObject> _Queue;
 
-        public PipelineReader(VisitTargetObject input, InputFileInfo[] inputPath)
+        public PipelineReader(VisitTargetObject input, InputFileInfo[] inputPath, PathFilter inputFilter)
         {
             _Input = input;
             _InputPath = inputPath;
+            _InputFilter = inputFilter;
             _Queue = new ConcurrentQueue<TargetObject>();
         }
 
@@ -33,8 +35,7 @@ namespace PSRule.Pipeline
             var targetObject = new TargetObject(sourceObject);
             if (_Input == null || skipExpansion)
             {
-                targetObject.Value.ConvertTargetInfoProperty();
-                _Queue.Enqueue(targetObject);
+                EnqueueInternal(targetObject);
                 return;
             }
 
@@ -44,10 +45,7 @@ namespace PSRule.Pipeline
                 return;
 
             foreach (var item in input)
-            {
-                targetObject.Value.ConvertTargetInfoProperty();
-                _Queue.Enqueue(item);
-            }
+                EnqueueInternal(item);
         }
 
         public bool TryDequeue(out TargetObject sourceObject)
@@ -72,6 +70,23 @@ namespace PSRule.Pipeline
                     Enqueue(PSObject.AsPSObject(_InputPath[i]));
                 }
             }
+        }
+
+        private void EnqueueInternal(TargetObject targetObject)
+        {
+            if (ShouldQueue(targetObject))
+                _Queue.Enqueue(targetObject);
+        }
+
+        private bool ShouldQueue(TargetObject targetObject)
+        {
+            if (_InputFilter != null && targetObject.Source != null && targetObject.Source.Count > 0)
+            {
+                foreach (var source in targetObject.Source.GetSourceInfo())
+                    if (!_InputFilter.Match(source.File))
+                        return false;
+            }
+            return true;
         }
     }
 }

--- a/src/PSRule/Pipeline/TargetObject.cs
+++ b/src/PSRule/Pipeline/TargetObject.cs
@@ -86,6 +86,7 @@ namespace PSRule.Pipeline
                 result.Add(targetInfo.Source);
 
             Value.ConvertTargetInfoProperty();
+            Value.ConvertTargetInfoType();
             result.AddRange(Value.GetSourceInfo());
             return result;
         }

--- a/tests/PSRule.Tests/PSRule.Options.Tests.ps1
+++ b/tests/PSRule.Tests/PSRule.Options.Tests.ps1
@@ -967,6 +967,45 @@ Describe 'New-PSRuleOption' -Tag 'Option','New-PSRuleOption' {
         }
     }
 
+    Context 'Read Input.IgnoreObjectSource' {
+        It 'from default' {
+            $option = New-PSRuleOption -Default;
+            $option.Input.IgnoreObjectSource | Should -Be $False;
+        }
+
+        It 'from Hashtable' {
+            $option = New-PSRuleOption -Option @{ 'Input.IgnoreObjectSource' = $True };
+            $option.Input.IgnoreObjectSource | Should -Be $True;
+        }
+
+        It 'from YAML' {
+            $option = New-PSRuleOption -Option (Join-Path -Path $here -ChildPath 'PSRule.Tests.yml');
+            $option.Input.IgnoreObjectSource | Should -Be $True;
+        }
+
+        It 'from Environment' {
+            try {
+                # With bool
+                $Env:PSRULE_INPUT_IGNOREOBJECTSOURCE = 'true';
+                $option = New-PSRuleOption;
+                $option.Input.IgnoreObjectSource | Should -Be $True;
+
+                # With int
+                $Env:PSRULE_INPUT_IGNOREOBJECTSOURCE = '1';
+                $option = New-PSRuleOption;
+                $option.Input.IgnoreObjectSource | Should -Be $True;
+            }
+            finally {
+                Remove-Item 'Env:PSRULE_INPUT_IGNOREOBJECTSOURCE' -Force;
+            }
+        }
+
+        It 'from parameter' {
+            $option = New-PSRuleOption -InputIgnoreObjectSource $True -Path $emptyOptionsFilePath;
+            $option.Input.IgnoreObjectSource | Should -Be $True;
+        }
+    }
+
     Context 'Read Input.IgnoreRepositoryCommon' {
         It 'from default' {
             $option = New-PSRuleOption -Default;

--- a/tests/PSRule.Tests/PSRule.Tests.yml
+++ b/tests/PSRule.Tests/PSRule.Tests.yml
@@ -59,6 +59,7 @@ execution:
 input:
   format: Yaml
   ignoreGitPath: false
+  ignoreObjectSource: true
   ignoreRepositoryCommon: false
   objectPath: items
   pathIgnore:

--- a/tests/PSRule.Tests/PathFilterTests.cs
+++ b/tests/PSRule.Tests/PathFilterTests.cs
@@ -1,4 +1,4 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 using System;
@@ -39,7 +39,9 @@ namespace PSRule
                 "**/obj/",
                 "",
                 "# Add reports/bin",
-                "!reports/bin/"
+                "!reports/bin/",
+                "**/ObjectFromFile*.json",
+                "!**/ObjectFromFile.json"
             };
             filter = PathFilter.Create(GetWorkingPath(), expressions, true);
             Assert.True(filter.Match("out/example.parameters.json"));
@@ -48,6 +50,8 @@ namespace PSRule
             Assert.True(filter.Match("src/bin/pwsh.exe"));
             Assert.True(filter.Match("src/obj/example.parameters.json"));
             Assert.False(filter.Match("reports/bin/other.json"));
+            Assert.False(filter.Match("ObjectFromFile.json"));
+            Assert.True(filter.Match("ObjectFromFileSingle.json"));
 
             // Exclude
             filter = PathFilter.Create(GetWorkingPath(), expressions, false);
@@ -87,9 +91,13 @@ namespace PSRule
             Assert.True(actual3.Match(".github/dependabot.yml"));
         }
 
+        #region Helper methods
+
         private static string GetWorkingPath()
         {
             return AppDomain.CurrentDomain.BaseDirectory;
         }
+
+        #endregion Helper methods
     }
 }

--- a/tests/PSRule.Tests/PipelineTests.cs
+++ b/tests/PSRule.Tests/PipelineTests.cs
@@ -58,7 +58,7 @@ namespace PSRule
             PSRuleOption.UseCurrentCulture(CultureInfo.InvariantCulture);
             var context = PipelineContext.New(GetOption(), null, null, null, null, null, new OptionContext(), null);
             var writer = new TestWriter(GetOption());
-            var pipeline = new GetRulePipeline(context, GetSource(), new PipelineReader(null, null), writer, false);
+            var pipeline = new GetRulePipeline(context, GetSource(), new PipelineReader(null, null, null), writer, false);
             try
             {
                 pipeline.Begin();
@@ -80,7 +80,7 @@ namespace PSRule
             option.Execution.InvariantCultureWarning = false;
             var context = PipelineContext.New(option, null, null, null, null, null, new OptionContext(), null);
             var writer = new TestWriter(option);
-            var pipeline = new GetRulePipeline(context, GetSource(), new PipelineReader(null, null), writer, false);
+            var pipeline = new GetRulePipeline(context, GetSource(), new PipelineReader(null, null, null), writer, false);
             try
             {
                 pipeline.Begin();
@@ -118,16 +118,44 @@ namespace PSRule
         {
             var option = GetOption();
             option.Rule.Include = new string[] { "FromFile1" };
-            var builder = PipelineBuilder.Invoke(GetSource(), option, null, null);
-            builder.InputPath(new string[] { "./**/ObjectFromFile.json" });
+            option.Input.PathIgnore = new string[]
+            {
+                "**/ObjectFromFile*.json",
+                "!**/ObjectFromFile.json"
+            };
+
+            // Default
             var writer = new TestWriter(option);
+            var builder = PipelineBuilder.Invoke(GetSource(), option, null, null);
+            builder.InputPath(new string[] { "./**/ObjectFromFile*.json" });
             var pipeline = builder.Build(writer);
             Assert.NotNull(pipeline);
             pipeline.Begin();
-            pipeline.Process(null);
+            pipeline.Process(GetTestObject());
+            pipeline.Process(GetFileObject());
             pipeline.End();
 
             var items = writer.Output.OfType<InvokeResult>().SelectMany(i => i.AsRecord()).ToArray();
+            Assert.Equal(4, items.Length);
+            Assert.True(items[0].HasSource());
+            Assert.True(items[1].HasSource());
+            Assert.True(items[2].HasSource());
+            Assert.True(items[3].HasSource());
+
+            // With IgnoreObjectSource
+            option.Input.IgnoreObjectSource = true;
+            writer = new TestWriter(option);
+            builder = PipelineBuilder.Invoke(GetSource(), option, null, null);
+            PipelineBuilder.Invoke(GetSource(), option, null, null);
+            builder.InputPath(new string[] { "./**/ObjectFromFile*.json" });
+            pipeline = builder.Build(writer);
+            Assert.NotNull(pipeline);
+            pipeline.Begin();
+            pipeline.Process(GetTestObject());
+            pipeline.Process(GetFileObject());
+            pipeline.End();
+
+            items = writer.Output.OfType<InvokeResult>().SelectMany(i => i.AsRecord()).ToArray();
             Assert.Equal(2, items.Length);
             Assert.True(items[0].HasSource());
             Assert.True(items[1].HasSource());
@@ -173,6 +201,25 @@ namespace PSRule
         private static string GetSourcePath(string fileName)
         {
             return Path.Combine(AppDomain.CurrentDomain.BaseDirectory, fileName);
+        }
+
+        private static PSObject GetTestObject()
+        {
+            var info = new PSObject();
+            var source = new PSObject();
+            source.Properties.Add(new PSNoteProperty("file", PSRuleOption.GetRootedPath("./ObjectFromFileNotFile.json")));
+            source.Properties.Add(new PSNoteProperty("type", "example"));
+            info.Properties.Add(new PSNoteProperty("source", new PSObject[] { source }));
+            var o = new PSObject();
+            o.Properties.Add(new PSNoteProperty("name", "FromObjectTest"));
+            o.Properties.Add(new PSNoteProperty("_PSRule", info));
+            return o;
+        }
+
+        private static PSObject GetFileObject()
+        {
+            var info = new FileInfo(PSRuleOption.GetRootedPath("./ObjectFromFileSingle.json"));
+            return new PSObject(info);
         }
 
         #endregion Helper methods

